src/frontend/src/optimizer/mod.rs

```diff
@@ -1118,6 +1120,38 @@ fn require_additional_exchange_on_root_in_local_mode(plan: PlanRef) -> bool {
         || exist_and_no_exchange_before(&plan, is_insert)
 }
 
+/// Result when applying an optimization rule.
+pub enum OResult<T> {
```

Comment:
What about putting this next to `trait Rule`? And for the name, IMHO `RuleResult` can be more accurate than `OResult`?